Reporting & replaying failing random seeds

funemy opened the pull request:
- When a failing case is found by `RandomScheduler`, shuttle now reports the random seed being used in addition to the failing schedule.
    - This requires a minor change in the implementation of `RandomScheduler`. Now, on a new execution, `RandomScheduler.rng` will be reinitialized with the new seed generated from `data_source`. 
       This is necessary because the two RNGs used in `RandomScheduler` (i.e., `rng` and `data_source`), though initialized using the same seed, advances at different paces. Therefore, the two RNGs will most likely be out of sync after the first iteration, and we cannot replay a failure case with just one seed from then. This change synchronizes the two RNGs at each iteration.
       This change does not seem to affect the performance of `RnadomScheduler`.
- Support replaying the failure with just random seeds.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jorajeev reviewed (commented):
Thanks for the PR.  Two major comments:

(1) The PCT scheduler is also initialized with a random seed - could you extend it in the same way you've extended the RandomScheduler

(2) Could you add a `check_random_with_seed` function that initializes the RandomScheduler with a given seed (and similarly add a `check_pct_with_seed`).  This would be useful when using Shuttle with a library like proptest which would generate the initial random seed.

funemy commented on src/lib.rs:
```diff
@@ -414,6 +414,49 @@ where
+/// Run the given function using random scheduler according to a seed.
+///
+/// This function allows deterministic replay of a failing schedule, as long as `f` contains no
+/// non-determinism other than that introduced by scheduling.
```
I updated the comment to make it clearer.

funemy commented:
@jorajeev I've updated the PR according to our discussion, ready for another review:

- Added `check_random_with_seed`
- Updated the comments about uncontrolled non-determinism
- Leaving PCT as it is for now.

jamesbornholt commented:
So the reason we didn’t implement replay this way the first time is that you lose the ability to detect divergence. If you’re debugging a failure from a schedule and insert some new code that happens to perform a synchronization operation, you’ll just silently start executing a different schedule from that point forward. The current version of replay is fairly likely to detect this kind of divergence.

I don’t have the context for why you want this change, but I’d like to make sure we don’t give up that protection in the common case. (And actually I’d love if we could make it even stronger, because it’s not guaranteed at the moment)

jorajeev commented:
> So the reason we didn’t implement replay this way the first time is that you lose the ability to detect divergence. If you’re debugging a failure from a schedule and insert some new code that happens to perform a synchronization operation, you’ll just silently start executing a different schedule from that point forward. The current version of replay is fairly likely to detect this kind of divergence.
> 
> I don’t have the context for why you want this change, but I’d like to make sure we don’t give up that protection in the common case. (And actually I’d love if we could make it even stronger, because it’s not guaranteed at the moment)

The need for this replay arises because we often run Shuttle in build pipelines where saving a failing schedule to a file is not easy.  (And printing to console is not feasible either, because build logs can get truncated.). For such situations, it's a lot more convenient to just print the single random seed that can be used to reproduce the failure locally.

Note that we're adding a new replay function, so the old one is still available (and the default if you use `check_random` as before).

You raise a good point about strengthening the check to avoid replay confusion when code changes.  I expect we could store additional information (maybe vector clocks, in some compressed form) to achieve this.

jorajeev commented on src/lib.rs:
```diff
@@ -333,6 +333,21 @@ where
+/// This effectively makes generating the random seed for each execution independent from
+/// `RandomScheduler`. Therefore, this can be used with a library (like proptest) that takes care of
+/// generating the random seeds.
+pub fn check_random_with_seed<F>(f: F, seed: u64)
```
This needs to take the number of iterations as a parameter, so it should be:

`pub fn check_random_with_seed<F>(f: F, seed: usize, iterations: usize)`

funemy replied:
updated.

sarsko commented on src/lib.rs:
```diff
@@ -333,6 +333,21 @@ where
     runner.run(f);
 }
 
+/// Run the given function under a random seed for one iteration.
```
It isn't clear what `under a random seed for one iteration` means. Also the `one iteration` part is wrong. We run the function `f` under a `RandomScheduler` initialized with the provided `seed` for `iterations` iterations.

funemy replied:
oops, good catch -- I forgot to update the comment

sarsko commented on src/lib.rs:
```diff
@@ -333,6 +333,21 @@ where
 }
 
+/// Run the given function under a random seed for one iteration.
+/// This effectively makes generating the random seed for each execution independent from
```
Nit: remove `effectively`

sarsko commented on src/lib.rs:
```diff
@@ -414,6 +429,52 @@ where
     runner.run(f);
 }
 
+/// Run the given function using random scheduler according to a seed.
```
Nit: `according to a seed` -> `initialized with the provided seed` (or similar)

sarsko commented on src/lib.rs:
```diff
@@ -414,6 +429,52 @@ where
+/// This is a convenience function for constructing a [`Runner`] that uses
+/// [`RandomScheduler::new_from_seed`](scheduler::RandomScheduler::new_from_seed) for one
+/// iteration.
+pub fn replay_random_from_seed<F>(f: F, seed: u64)
```
Not sure there is any point in having this function. This is just `check_random_with_seed(f, seed, 1)`.

sarsko commented on src/scheduler/random.rs:
```diff
@@ -18,6 +18,33 @@ pub struct RandomScheduler {
+    fn drop(&mut self) {
+        if let Some(s) = self.inner {
+            eprintln!(
+                "failing seed:\n\"\n{}\n\"\npass the seed to `shuttle::replay_from_seed` to replay the failure.",
```
If we decide to do replay based on environment variables, then the comment here needs to be updated.

sarsko commented on tests/mod.rs:
```diff
@@ -94,6 +94,25 @@ where
     // `FailurePersistence`s for each test
 }
 
+/// Validates that replay-from-seed works by running a failing seed found by a random scheduler for one
```
Nit: `replay-from-seed` isn't anything. Either name the function (`replay_random_from_seed`) or "abstract over it"; "the replay from seed functionality".

sarsko commented on tests/mod.rs:
```diff
@@ -94,6 +94,25 @@ where
+        })
+        .expect_err("replay should panic")
+    };
+    let new_output = result.downcast::<String>().unwrap();
```
Nit: `new_output`: there is no other `output`.

sarsko commented on tests/mod.rs:
```diff
@@ -94,6 +94,25 @@ where
+        .expect_err("replay should panic")
+    };
+    let new_output = result.downcast::<String>().unwrap();
+    let new_schedule = parse_schedule::from_stdout(&new_output).expect("output should contain a schedule");
```
nit: `new_schedule` is non-descriptive. Maybe `schedule_from_replay` ? Maybe rename `schedule` in the params to `expected_schedule` ?

funemy commented:
@sarsko all comments should be addressed, the big changes are:

- `reply_random` and `reply_random_from_seed` are removed
- `check_random` now first tries reading from the env var `SHUTTLE_REPLAY_SEED` and if a valid seed is read, initializes the scheduler with it
- error messages printed by `CurrentSeedDropGuard` is updated.

sarsko commented on src/lib.rs:
```diff
@@ -322,13 +322,46 @@ where
     use crate::scheduler::RandomScheduler;
 
-    let scheduler = RandomScheduler::new(iterations);
+    let seed_env = std::env::var("SHUTTLE_REPLAY_SEED");
```
Not sure how much I care, but calling it `SHUTTLE_REPLAY_SEED` is odd, as it strictly speaking is independent of replaying.

funemy replied:
maybe `SHUTTLE_RANDOM_SEED`?

jorajeev replied:
Yeah, given the changes, let's go with `SHUTTLE_RANDOM_SEED`

sarsko reviewed (commented):
Looks good to me, a couple of minor comments. Would like one more look from Rajeev/James/Bernhard.

jorajeev commented on src/lib.rs:
```diff
@@ -322,13 +322,46 @@ where
 
-    let scheduler = RandomScheduler::new(iterations);
+    let seed_env = std::env::var("SHUTTLE_REPLAY_SEED");
+    let runner;
```
The idiomatic way would be to write

`let runner = match seed_env { .. };`

jorajeev commented on src/lib.rs:
```diff
@@ -322,13 +322,46 @@ where
+                println!(
+                    "Initializing RandomScheduler with the seed provided by SHUTTLE_REPLAY_SEED: {}",
+                    seed
+                );
```
+1

jorajeev commented on src/scheduler/random.rs:
```diff
@@ -18,6 +18,33 @@ pub struct RandomScheduler {
+}
+
+impl CurrentSeedDropGuard {
+    fn wipe(&mut self) {
```
Let's call this `clear` (or `reset`)  instead of `wipe`

funemy commented:
@sarsko @jorajeev all comments addressed
